File: Client/src/app/core/services/course.service.ts
import { Injectable } from '@angular/core';
import { ApiPaths } from '@core/constants/api-paths';
import { CourseDTO } from '@core/models/course.model';
import { Observable, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { ApiService } from './api.service';

@Injectable({
  providedIn: 'root',
})
/**
 * Service for managing course-related operations.
 * This service provides methods to fetch, create, update, and delete courses,
 * as well as to fetch entities associated with a course.
 * @export
 * @class CourseService
 * * @implements {ICourseService}
 * * @see {@link ICourseService} for the interface definition.
 * */
export class CourseService {
  constructor(private apiService: ApiService) {}

  // ================================================================
  // CRUD operations
  // ================================================================

  /**
   * Fetches all courses from the API.
   * @returns An observable of an array of CourseDTO.
   */
  getCourses(): Observable<CourseDTO[]> {
    return this.apiService.get<CourseDTO[]>(ApiPaths.COURSES).pipe(
      catchError((error) => {
        console.error('CourseService: Error fetching courses', error);
        return throwError(() => new Error('Failed to load courses'));
      })
    );
  }

  /**
   * Fetches a specific course by code.
   * @param courseCode The code of the course to fetch.
   * @returns An observable of the CourseDTO.
   */
  getCourse(courseCode: string): Observable<CourseDTO> {
    return this.apiService
      .get<CourseDTO>(`${ApiPaths.COURSES}/${courseCode}`)
      .pipe(
        catchError((error) => {
          console.error('CourseService: Error fetching course', error);
          return throwError(() => new Error('Failed to load course'));
        })
      );
  }

  /**
   * Creates a new course.
   * @param course The course data to create.
   * @returns An observable of the created CourseDTO.
   */
  createCourse(course: Partial<CourseDTO>): Observable<CourseDTO> {
    return this.apiService.post<CourseDTO>(ApiPaths.COURSES, course).pipe(
      catchError((error) => {
        console.error('CourseService: Error creating course', error);
        return throwError(() => new Error('Failed to create course'));
      })
    );
  }

  /**
   * Updates an existing course.
   * @param courseCode The code of the course to update.
   * @param course The updated course data.
   * @returns An observable of the updated CourseDTO.
   */
  updateCourse(
    courseCode: string,
    course: Partial<CourseDTO>
  ): Observable<CourseDTO> {
    return this.apiService
      .put<CourseDTO>(`${ApiPaths.COURSES}/${courseCode}`, course)
      .pipe(
        catchError((error) => {
          console.error('CourseService: Error updating course', error);
          return throwError(() => new Error('Failed to update course'));
        })
      );
  }

  /**
   * Deletes a course by code.
   * @param courseCode The code of the course to delete.
   * @returns An observable that completes when the deletion is successful.
   */
  deleteCourse(courseCode: string): Observable<void> {
    return this.apiService
      .delete<void>(`${ApiPaths.COURSES}/${courseCode}`)
      .pipe(
        catchError((error) => {
          console.error('CourseService: Error deleting course', error);
          return throwError(() => new Error('Failed to delete course'));
        })
      );
  }

  // ================================================================
  // Business logic operations
  // ================================================================
}

File: Server/src/main/java/com/egabi/university/service/academic/CourseService.java
package com.egabi.university.service.academic;

import com.egabi.university.dto.CourseDTO;

import java.util.List;

public interface CourseService {
    // ================================================================
    // CRUD Methods
    // ================================================================
    
    /**
     * Get all courses available in the system.
     *
     * @return a list of all courses as CourseDTO objects
     */
    List<CourseDTO> getAllCourses();
    
    /**
     * Get Course by code.
     *
     * @param code the code of the course
     * @return Course object if exists, null otherwise
     */
    CourseDTO getCourseByCode(String code);
    
    /**
     * Create a new course.
     *
     * @param courseDTO the course data transfer object containing course details
     * @return the created CourseDTO object
     */
    CourseDTO createCourse(CourseDTO courseDTO);
    
    /**
     * Update an existing course.
     *
     * @param code      the code of the course to be updated
     * @param courseDTO the course data transfer object containing updated course details
     * @return the updated CourseDTO object
     */
    CourseDTO updateCourse(String code, CourseDTO courseDTO);
    
    /**
     * Delete a course by its code.
     *
     * @param code the code of the course to be deleted
     */
    void deleteCourse(String code);
    
    // ================================================================
    // Business Logic Methods
    // ================================================================
    
    /**
     * Get all courses by department ID.
     *
     * @param departmentId the ID of the department
     * @return a list of CourseDTO objects associated with the specified department
     */
    List<CourseDTO> getCoursesByDepartmentId(Long departmentId);
    
    /**
     * Count all courses for a specific department.
     *
     * @param departmentId the ID of the department
     * @return the count of courses associated with the specified department
     */
    Long countCoursesByDepartmentId(Long departmentId);
}
